### PR TITLE
OKRs - 2019 Q4 - Package Managers

### PR DESCRIPTION
**Pull request by Stebalien**
Rough draft of some OKRs to get us started. I focused on motivation/discussion instead of trying to nail down the exact KRs.

Resourcing:

* I'm assuming that @achingbrain will still be working on the async await refactor for much of the quarter. Unfortunately, that's not something we can reasonably drop.

### CODE REVIEW

**Comment by momack2** [comment]
Ref https://github.com/ipfs/team-mgmt/issues/1033

File: OKR/PACKAGE-MANAGERS.md [hunk 1]
```diff
@@ -2,6 +2,224 @@
...
+
+Ship a blog post about speeding up CI by using go-ipfs to distribute docker images.
+
+(docker _is_ a package manager, or close enough)
+
+##### Demo: Demonstrate performance improvements adding NPM to IPFS
+
+**Owner** @aschmahmann/@achingbrain
+
+Ship a demo of adding NPM to IPFS.
```
Comment (momack2) [inline OKR/PACKAGE-MANAGERS.md]: other ideas: time to mirror ipfs-wikipedia (a data package made up of many small files), apt-on-ipfs and clojars-on-ipfs will really focus the work here

File: OKR/PACKAGE-MANAGERS.md [hunk 2]
```diff
@@ -2,6 +2,224 @@
...
+
+Roadmap:
+
+1. Testground is up and running and can test a >10K node DHT.
+2. Proposed DHT fixes are tested against various network scenarios and we are confident they will significantly improve the state of the network.
+3. We cut a release.
+
+##### P0: Fix the DHT
+
+@aschmahmann will likely be pulled into working on DHT fixes during the second half of the quarter (once we can actually start _testing_ DHT fixes).
```
Comment (momack2) [inline OKR/PACKAGE-MANAGERS.md]: @aschmahmann - sound good to you?

File: OKR/PACKAGE-MANAGERS.md [hunk 3]
```diff
@@ -2,6 +2,224 @@
...
+Improve block prefetching when downloading a file so we can actually saturate bitswap.
+
+##### P2: Add/MFS
+
+**KR**: Add
+**Owner:** @aschmahmann (?)
+
+`ipfs add` needs to be fast. Given our exploration during Q3, it looks like we'll need to improve the performance of go-mfs to achieve this.
+
+##### P3: Datastore Performance
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply OKR/PACKAGE-MANAGERS.md]: I didn't factor him in as a resource but if he's interested, that would be great!

File: OKR/PACKAGE-MANAGERS.md [hunk 4]
```diff
@@ -2,6 +2,224 @@
...
+
+#### Alternatives
+
+These are some alternative initiatives that _could_ have addressed the KRs.
+
+* Add/Update: A stand-alone IPFS "dropbox" tool. We could write an entirely new tool that simply makes a directory available to the IPFS network using the filestore. Files would be automatically added/removed from IPFS as they're added/removed from the directory.
+  * Pro: Many users would love this kind of tool.
+  * Con: Could be a _lot_ of work.
+  * Con: We want mount support and metadata _anyways_.
+* Update: Out-of-band metadata: Instead of storing timestamps in files to enable rsync, we could store them out-of-band.
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply OKR/PACKAGE-MANAGERS.md]: We did. This section just lists alternatives I considered (at one point, at least) so we have a record of the pros/cons.

File: OKR/PACKAGE-MANAGERS.md [hunk 5]
```diff
@@ -2,6 +2,224 @@
...
+* Updating from one snapshot to the next.
+* Mounting the repository locally and having the package maintenance tools operate on it directly.
+
+##### P1: Fetch
+
+IPFS can quickly and efficiently fetch packages from a large package repository.
+
+TODO: define "quickly and efficiently" in terms of beating HTTP.
+
+TODO: consider breaking this into two KRs? In the past, we had one for efficiency and one for throughput.
```
Comment (Stebalien) [inline OKR/PACKAGE-MANAGERS.md]: 1. IPFS uses at most 10% more bandwidth than HTTP.
2. IPFS maxes out a 1GiB link on the internet (100ms latencies).

File: OKR/PACKAGE-MANAGERS.md [hunk 6]
```diff
@@ -2,6 +2,224 @@
...
+
+* IPNS: To build a fully decentralized package manager, IPNS is a must. However, to get in the door with existing package managers, IPFS just needs to be a viable _transport_.
+* Performant GC: Package maintainers often keep old packages around (TODO: check this) so GC isn't all that important. Package manager _users_ can throw away their local repo when it grows too large (i.e., treat it like a cache).
+* Pinning Usability (named pins, etc.): While a pain-point and embarrassing UX issue, pinning and pin management is not a show-stopper for package managers.
+
+##### P0: Update (P0)
+
+IPFS can quickly update a large package repository.
+
+TODO: "quickly" should be defined in terms of normal package update procedures used by package maintainers for some package manager. This could be:
```
Comment (Stebalien) [inline OKR/PACKAGE-MANAGERS.md]: https://github.com/ipfs/package-managers/issues/18#issuecomment-471314183

File: OKR/PACKAGE-MANAGERS.md [hunk 7]
```diff
@@ -2,6 +2,224 @@
...
+
+TODO: define "quickly and efficiently" in terms of beating HTTP.
+
+TODO: consider breaking this into two KRs? In the past, we had one for efficiency and one for throughput.
+
+##### P2: Add
+
+IPFS can quickly add a large package repository.
+
+TODO: "quickly" should be defined in terms of adding some well-known package repo or test dataset.
```
Comment (Stebalien) [inline OKR/PACKAGE-MANAGERS.md]: https://github.com/ipfs/package-managers/issues/18#issue-418828188

File: OKR/PACKAGE-MANAGERS.md (hunk 2 above)
Comment (aschmahmann) [reply OKR/PACKAGE-MANAGERS.md]: Works for me, seems pretty important since we need it to get a release out.

File: OKR/PACKAGE-MANAGERS.md (hunk 5 above)
Comment (aschmahmann) [reply OKR/PACKAGE-MANAGERS.md]: I assume we already know this, but was at the ICN conference this past week and a paper was presented (https://conferences.sigcomm.org/acm-icn/2019/proceedings/icn19-34.pdf) that shows (in go-ipfs 0.4.20) a lot of over requesting of data from Bitswap on a small closed network of 10 nodes.

Is the 10% you mention between 2 peers, or even in a multi-peer environment?

File: OKR/PACKAGE-MANAGERS.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (hsanjuan) [reply OKR/PACKAGE-MANAGERS.md]: I cannot make promises

File: OKR/PACKAGE-MANAGERS.md (hunk 1 above)
Comment (aschmahmann) [reply OKR/PACKAGE-MANAGERS.md]: I'm fine with whatever demo we decide is most effective. As I mentioned in my IPNS comment above we should figure out what the main "this is cool" factors are aside from it just working on IPFS.

As some of our research docs https://github.com/ipfs/package-managers/tree/master/docs (including https://app.mural.co/t/protocollabs6957/m/protocollabs6957/1557168696127/577c9453a3c51199c8163cf0fe5701294e55f99b) indicate there are a number of things we could focus on. They might be things like faster downloads of packages, or ensuring that specific package versions cannot be deleted. Adding in offline operation might be an interesting thing to demonstrate as well.

File: OKR/PACKAGE-MANAGERS.md (hunk 5 above)
Comment (momack2) [reply OKR/PACKAGE-MANAGERS.md]: I think steven's comment is the KR goal - not what we currently hit...

File: OKR/PACKAGE-MANAGERS.md [hunk 8]
```diff
@@ -2,6 +2,224 @@
...
+
+##### P3: Communicate add performance improvements
+
+Communicate add performance improvements to the community.
+
+#### Initiatives
+
+In each of these initiatives, we plan to pair with a user/community member.
+
+##### Demo: OpenSuse + Mount
```
Comment (momack2) [inline OKR/PACKAGE-MANAGERS.md]: Question whether RPM is the right tool here - if use rsync likely a good example of where we can get the benefit. Arch sufficiently small that fits nicely on a hard drive...

Comment (momack2) [reply OKR/PACKAGE-MANAGERS.md]: @djdv for thoughts?

File: OKR/PACKAGE-MANAGERS.md [hunk 9]
```diff
@@ -2,6 +2,224 @@
...
+
+##### P1: Communicate fetch performance improvements
+
+Communicate fetch performance improvements to the community.
+
+##### P3: Communicate add performance improvements
+
+Communicate add performance improvements to the community.
+
+#### Initiatives
```
Comment (momack2) [inline OKR/PACKAGE-MANAGERS.md]: we can change the initiative topic at any point - goal is to make the demo =]

File: OKR/PACKAGE-MANAGERS.md (hunk 8 above)
Comment (djdv) [reply OKR/PACKAGE-MANAGERS.md]: During a meeting today we discussed the expected trajectory.
We'll aim for openSUSE/RPM but may or may not experiment with other repos/formats as a means of ramping up. Starting with locally made smaller repos for practicality and working up to mirroring actual (large) sets.

Going from 1 target to another seems less reasonable so we'll likely remain with RPM the whole way through.

File: OKR/PACKAGE-MANAGERS.md (hunk 1 above)
Comment (Stebalien) [reply OKR/PACKAGE-MANAGERS.md]: The goal of this initiative is to demo the performance improvements to `ipfs add`.

File: OKR/PACKAGE-MANAGERS.md [hunk 10]
```diff
@@ -2,6 +2,224 @@
...
+
+While package repositories are large, they tend not to change _all_ that much on a daily basis. We can side-step the add performance issues by ensuring that package maintainers can easily and efficiently _update_ package repositories stored in IPFS.
+
+Finally, while we've been focusing on how to ensure IPFS meets the needs of package _maintainers_, we've neglected the end user. It's unlikely that a package manager would choose to _exclusively_ use IPFS and end-users won't put up with IPFS unless it's at least as fast as HTTP and reasonably efficient. This came up when one of our partners found that using IPFS to distribute docker images within a fast datacenter network was no faster than going directly to docker-hub.
+
+**Not Included**
+
+* IPNS: To build a fully decentralized package manager, IPNS is a must. However, to get in the door with existing package managers, IPFS just needs to be a viable _transport_.
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply OKR/PACKAGE-MANAGERS.md]: > I don't think I buy this. DNSLink is insufficient if users care about 1) Update speed 2) Offline operation or 3) Third party persistence (i.e. not relying on the publisher to continue to maintain their package)

The goal is table-stakes.

1. System package managers often have multi-hour-to-day delays between releasing a package and mirrors updating. Nobody's going to notice the DNS propagation time (which is often on the order of minutes).
2. Table stakes. Offline operation is _great_, but we're just trying to be _viable_.
3. Again, table-stakes.

At this point, we're not shooting for a compelling user story. We're trying to make IPFS _viable_. At the moment.

1. Adding can take _days_.
2. Updating isn't _possible_ (without a bunch of custom tooling or re-adding).
3. Downloading is extremely slow (it rarely gates faster than a few megabytes per second) and has an insane overhead (at least +50%, usually more like 100%).

Given that and our resource constraints, we have to make a choice. We can't choose implementing a compelling user story over fixing an issue that prevents users from even considering IPFS.

**Comment by momack2** [comment]
I'm going to merge this as is because it has a ton of really useful rhetoric outside the OKR sheet

**Pull request merged**